DotVVM.Diagnostics.StatusPage/DotHtmlFilesRuntimePrecompiler.cs

```diff
@@ -0,0 +1,180 @@
+                    .Select(s => new DotHtmlFileInfo()
+                    {
+                        TagName = s.TagName,
+                        VirtualPath = s.Src,
+                        Namespace = s.Namespace,
+                        Assembly = s.Assembly,
+                        TagPrefix = s.TagPrefix
+                    }).ToList());
+            }
+        }
+
+        internal void BuildView(DotHtmlFileInfo file, ConcurrentBag<DotHtmlFileInfo> tempList)
+        {
+            Debug.WriteLine($"Precompiling path: {file?.VirtualPath}");
+            if (file.Status != CompilationState.NonCompilable)
+            {
+                try
+                {
+                    var controlFactory = serviceProvider.GetRequiredService<IControlBuilderFactory>();
+
+                    var pageBuilder = controlFactory.GetControlBuilder(file.VirtualPath);
+
+                    var compiledControl = pageBuilder.builder.Value.BuildControl(controlFactory, serviceProvider);
+
+                    if (compiledControl is DotvvmView view && view.Directives.TryGetValue(
+                            ParserConstants.MasterPageDirective,
+                            out var masterPage))
+                    {
+                        if (MasterPages.All(s => s.VirtualPath != masterPage) &&
+                            tempList.All(s => s.VirtualPath != masterPage))
+                        {
+                            tempList.Add(new DotHtmlFileInfo()
+                            {
+                                VirtualPath = masterPage
+                            });
+                        }
+                    }
+
+                    file.Status = CompilationState.CompletedSuccessfully;
+                    file.Exception = null;
+                }
+                catch (Exception e)
+                {
+                    file.Status = CompilationState.CompilationFailed;
+                    file.Exception = e.Message;
+                }
+            }
+        }
+        /// <summary>
+        /// Builds all dothtml file registered in configuration. When the files are already compiled and cached, compilation is not executed.  
+        /// </summary>
+        public async Task<bool> CompileAll()
+        {
+            ResolveAllFiles();
+            var tempMasterPages = new ConcurrentBag<DotHtmlFileInfo>();
+            await BuildPagesAndControls(tempMasterPages);
+            await BuildMasterPages(tempMasterPages);
+            return MasterPages.Concat(Controls).Concat(Routes).All(s => s.Status != CompilationState.CompilationFailed);
+        }
+
+        private async Task<ConcurrentBag<DotHtmlFileInfo>> BuildMasterPages(ConcurrentBag<DotHtmlFileInfo> tempMasterPages)
+        {
+            while (tempMasterPages.Count > 0)
+            {
+                tempMasterPages.ToList().ForEach(masterPages.Add);
+                masterPages = new ConcurrentBag<DotHtmlFileInfo>(masterPages.Distinct());
```

Comment:
Reassigning the `ConcurrentBag` does not seem like a great idea, since there may be multiple requests running in parrallel which will just lead to race conditions.

Reply:
Also the `DotHtmlFileInfo` does not even implement `Equals`+`GetHashCode`, so the Distinct call is not going to throw out duplicates.